app/services/search/tag.rb

```diff
@@ -0,0 +1,84 @@
+module Search
```

Comment:
Is there going to be a lot of variability between models when setting up indexes?

If so...skip everything I'm about to say and have a fantastic rest of your day - nothing left to see here 😂

If not...what if we made some base class for Search that these services can then inherit from?

The base class would have these methods implemented and we could overwrite them for a specific use case if needed. It would mean that most of these implementations are the `INDEX_NAME`, `INDEX_ALIAS`, settings, and mappings. So then `Tag < Search` and this file would have those things defined, the rest being inherited from Search.

Something like this?
```ruby
class Search
  class << self
    def get(id)
      SearchClient.get(id: id, index: self::INDEX_ALIAS)
    end

    def create_index(index_name = self::INDEX_NAME)
      SearchClient.indices.create(index: index_name, body: self.settings)
    end

    def delete_index(index_name = self::INDEX_NAME)
      SearchClient.indices.delete(index: index_name)
    end

    def add_alias(index_name = self::INDEX_NAME, index_alias = self::INDEX_ALIAS)
      SearchClient.indices.put_alias(index: index_name, name: index_alias)
    end

    def update_mappings(index_alias = self::INDEX_ALIAS)
      SearchClient.indices.put_mapping(index: index_alias, body: self.mappings)
    end
  end
end
```

```ruby
class Tag < Search
  INDEX_NAME = "tag_#{Rails.env}".freeze
  INDEX_ALIAS = "tag_#{Rails.env}_alias".freeze

  class << self
    # Overwriting the create_index method inherited from Search because it's slightly different
    def create_index(id, new_argument)
        SearchClient.indices.create(index: index_name, body: settings, new_argument: new_argument)
    end

    def settings
      { settings: { index: index_settings } }
    end

    def index_settings
      if Rails.env.production?
        {
          number_of_shards: 1,
          number_of_replicas: 1
        }
      else
        {
          number_of_shards: 1,
          number_of_replicas: 0
        }
      end
    end

    def mappings
      {
        dynamic: "strict",
        properties: {
          id: {
            type: "keyword"
          },
          name: {
            type: "text",
            fields: {
              raw: {
                type: "keyword"
              }
            }
          },
          hotness_score: {
            type: "integer"
          },
          supported: {
            type: "boolean"
          },
          short_summary: {
            type: "text"
          }
        }
      }
    end
  end
end
```
If you've read this far (🙈) I was thinking this might help DRY it up for when we add more services/models to ES in the future. Less chance of forgetting to include some of the "core" methods.

Reply:
@atsmith813 this is exactly where my head was going as well. At Kenna we had a bunch of query classes and they all included a searchable module. We questioned that a lot and if inheritance would have been a better way to go, I think it is. My plan is to start building and feel it out before making the abstraction to make sure it was the one we really want. First class, keep it like this, then second class abstract it. 95% sure we are headed that way but I am a big fan of holding off abstractions until you absolutely need them, that way you can ensure you are making the right ones. 

I'm thinking keep the namespace and use a Base class or something for it
```ruby
module Search
  class Base
    ....
  end
end
```

Reply:
Super happy with how this discussion is going. Yay for holding off on abstractions until we need them. 

In the past when using base classes like the one discussed here, I often opted for methods instead of constants:

```diff
- SearchClient.get(id: id, index: self::INDEX_ALIAS)
+ SearchClient.get(id: id, index: index_alias)
```

Then the base class would add a default implementation like

```ruby
def index_alias
  raise NotImplementedError
end
```

This made it easier to figure out when you forgot to add a needed constant/typoed it etc.